### PR TITLE
bosun: add annotation funcs to expr library

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

**Comment by kylebrandt** [comment]
Antable in action in Grafana:

image

File: cmd/bosun/expr/annotate.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+		F:      AnDurations,
+	},
+	"antable": {
+		Args:   []models.FuncType{models.TypeString, models.TypeString, models.TypeString, models.TypeString},
+		Return: models.TypeTable,
+		F:      AnTable,
+	},
+}
+
+func procDuration(e *State, startDuration, endDuration string) (*time.Time, *time.Time, error) {
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: I'd like a better name for this if possible. It is a bit ambiguous.

File: cmd/bosun/expr/annotate.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations := annotate.Annotations{}
+	for _, a := range annotations {
+		if filter == "" {
+			filteredAnnotations = append(filteredAnnotations, a)
+			continue
+		}
+		match, err := boolq.AskExpr(filter, a)
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: Is there a way to use boolq so that we only parse the filter once?

File: cmd/bosun/expr/annotate.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		if err != nil {
+			return nil, err
+		}
+		if match {
+			filteredAnnotations = append(filteredAnnotations, a)
+		}
+	}
+	sort.Sort(sort.Reverse(annotate.AnnotationsByStartID(filteredAnnotations)))
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: does the annotation server endpoint not sort them already?

File: cmd/bosun/expr/annotate.go (hunk 1 above)
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: also, why does this need to return pointers? If it is either an error, or valid times, just return `time.Time{}` in the error case. Saves some of our `*` budget.

File: cmd/bosun/expr/annotate.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	series := make(Series)
+	for i, a := range filteredAnnotations {
+		aStart := a.StartDate.Time
+		aEnd := a.EndDate.Time
+		inBounds := (aStart.After(*reqStart) || aStart == *reqStart) && (aEnd.Before(*reqEnd) || aEnd == *reqEnd)
+		entirelyOutOfBounds := aStart.Before(*reqStart) && aEnd.After(*reqEnd)
+		aDuration := aEnd.Sub(aStart)
+		if inBounds {
+			// time has no meaning here, so we just make the key an index since we don't have an array type
+			series[time.Unix(int64(i), 0).UTC()] = aDuration.Seconds()
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: I hate this misuse of timestamps. It is pretty much useless without some aggregation (I'm assuming sum) is applied on top. So why not aggregate here and return a number?

File: cmd/bosun/expr/annotate.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+		}
+		if entirelyOutOfBounds {
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: I'd feel more comfortable if this was an else if chain. Just because.

File: cmd/bosun/expr/annotate.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+			// Duration is equal to that of the full request
+			series[time.Unix(int64(i), 0).UTC()] = reqEnd.Sub(*reqStart).Seconds()
+		}
+		if aDuration == 0 {
+			// This would mean an out of bounds. Should never be here, but if we don't return an error in the case that we do end up here then we might panic on divide by zero later in the code
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: this means the time filter function is broken, no? I say silently skip it. But I guess this works too. nvm.

File: cmd/bosun/expr/annotate.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+	reqStart, reqEnd, err := procDuration(e, startDuration, endDuration)
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations, err := getAndFilterAnnotations(e, reqStart, reqEnd, filter)
+	if err != nil {
+		return nil, err
+	}
+	series := make(Series)
+	for i, a := range filteredAnnotations {
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: straight up duplication. Makes me sad. could potentially make a helper function that takes a `useFullDuration` flag.

at beginning set `value = (duration or 1)` by flag. And when adding to the series multiply that value by the percentage calculated in the window, and use that for the data point.

File: cmd/bosun/expr/annotate.go [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+	return &Results{
+		Results: []*Result{
+			{Value: t},
+		},
+	}, nil
+}
+
+// hhmmss formats a duration into HHH:MM:SS (Hours, Minutes, Seconds) so it can be lexically sorted
+// up to 999 hours
+func hhmmss(d time.Duration) string {
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: wouldn't that be `hhhmmss`?

File: cmd/bosun/expr/annotate.go [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+	columns := strings.Split(fieldsCSV, ",")
+	columnLen := len(columns)
+	if columnLen == 0 {
+		return nil, fmt.Errorf("must specify at least one column")
+	}
+	columnIndex := make(map[string]int, columnLen)
+	for i, v := range columns {
+		// switch is so we fail before fetching annotations
+		switch v {
+		case "start", "end", "owner", "user", "host", "category", "url", "message", "duration", "link":
```
Comment (captncraig) [inline cmd/bosun/expr/annotate.go]: kinda disliking this style. Could this be more data-driven? Thinking something like define a `map[string]func(annotation)interface{}` and all of the column checks and data fetching use that. Could kill the giant switch, and some of the duplication here. Maybe?

File: cmd/bosun/expr/annotate.go (hunk 4 above)
Comment (kylebrandt) [inline cmd/bosun/expr/annotate.go]: Same type that t() returns. Could make a new type but that is hard. The reason to return a series is that so you can:

 * See a list of the individual items
 * more importantly in the case of ancounts, len vs sum lets you distinguish between fractional outages vs the number of outages.

File: cmd/bosun/expr/annotate.go (hunk 1 above)
Comment (kylebrandt) [inline cmd/bosun/expr/annotate.go]: Thanks for catching this, this makes the rest of the code a lot cleaner.

**Pull request merged**